### PR TITLE
[#15] Get extra files

### PR DESCRIPTION
**Pull request by GuChad369**
This is just one step to start. I found that this function involves a lot of things and is not easy to achieve. I started this PR to help us discuss the details.

I want to describe my idea about this first.

I have created a function called `pgmoneta_ext_get_file(arg)` which returns the bytes of the file specified by the argument passed to it.

For the extension part, I prefer to create the following functions:
- `pgmoneta_ext_get_file(arg)`: This is the current version and it will return the bytes of the specified file passed in.
- `pgmoneta_ext_get_files(arg)`: This will return all file paths in the specified directory passed in.

For the core part, the process to get extra files is as follows:
- If we want to get a single file, we call `pgmoneta_ext_get_file` to get the bytes of the file and then pass it to the backend.
- If we want to get all files in a directory, we first use `pgmoneta_ext_get_files` to get all files in that directory. Then, we use a loop to pass all the files to the backend.

Achieving this function is not easy, and we may need to change it as we progress.

### CODE REVIEW

**Comment by GuChad369** [comment]
How should we decide the privilege of getting the file? Currently, we only allow the `superuser` to get files.

**Comment by jesperpedersen** [comment]
Well, PostgreSQL runs under an user account, so we should be able to use those credentials

**Comment by jesperpedersen** [comment]
Try and follow the

```sql
SELECT session_user, current_user;
```

path...

**Comment by GuChad369** [comment]
Since `pgmoneta` will use the `repl` role to access PostgreSQL, should we give the privilege to the `repl` role?

Or do we need to define a function to grant a role the privilege to access the file?

`superuser` can use this function to grant the privilege to one user.

**Comment by jesperpedersen** [comment]
There need to be options to for the `repl` role, and those will enable features in the extension.

So, a lot will be based on documentation for this

**Comment by jesperpedersen** [comment]
We can start with a `SUPERUSER` requirement, and then improve over time

**Comment by jesperpedersen** [comment]
So, from pgmoneta/core get the priviledge, and if it isn't `SUPERUSER` (to begin with), then you will log `ERROR` instead

**Comment by GuChad369** [comment]
Add a new step to cast binary data to characters to ensure we can handle this result on the core side. This will allow us to manage both text and binary files effectively.

**Comment by GuChad369** [comment]
These two functions can include a new feature to limit access to directories. We can add an array to list all directories that users can access, while other directories will not be accessible.

Alternatively, we can modify this later, incorporating user role limitations as well.

**Comment by jesperpedersen** [comment]
I'm not following - we should just try to retrieve the files requested. If it doesn't work then `NULL` should be returned.

And, we don't want to scan the entire machine for directories that can be accessed...

**Comment by GuChad369** [comment]
> I'm not following - we should just try to retrieve the files requested. If it doesn't work then `NULL` should be returned.

This is the current behavior of the function.

> And, we don't want to scan the entire machine for directories that can be accessed...

Since I saw requests for access outside the `/data` directory, I meant we can add a step to check if the file path is under `/data`. If it is, then return `NULL`.

**Comment by jesperpedersen** [comment]
I don't think you need to do that - maybe the user want an extra copy of `pg_control`

File: src/include/utils.h [hunk 1]
```diff
@@ -70,6 +76,14 @@ pgmoneta_ext_check_role(Oid roleid, const char* rolename);
 int
 pgmoneta_ext_check_privilege(Oid roleid);
 
+/**
+ * Recursively list files in the directory and accumulate their paths in the provided array state.
+ * @param name The directory or file name to check.
+ * @param astate The array build state to accumulate file paths.
+ */
+void
+list_files(const char* name, ArrayBuildState* astate);
```
Comment (jesperpedersen) [inline src/include/utils.h]: Please, prefix with `pgmoneta_ext_` to all functions in `.h` files

File: src/pgmoneta_ext/lib.c [hunk 2]
```diff
@@ -28,31 +28,53 @@
  */
 
 /* pgmoneta_ext */
-#include <pgmoneta_ext.h>
-#include <utils.h>
+#include "pgmoneta_ext.h"
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: Why the `"` ?

File: src/include/utils.h (hunk 1 above)
Comment (jesperpedersen) [reply src/include/utils.h]: Should this be `static` ?

File: src/pgmoneta_ext/utils.c [hunk 3]
```diff
@@ -28,18 +28,18 @@
  */
 
 /* pgmoneta_ext */
-#include <utils.h>
+#include "utils.h"
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/utils.c]: Again, `"`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
This doesn't compile with PostgreSQL 16 - test with 13->17, and add the functions to the manual

**Comment by jesperpedersen** [comment]
```
In file included from pgmoneta/pgmoneta_ext/src/include/utils.h:39,
                 from pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c:32:
pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c: In function ‘pgmoneta_ext_get_files’:
pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c:305:29: error: passing argument 1 of ‘PointerGetDatum’ makes pointer from integer without a cast [-Wint-conversion]
  305 |       PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate, CurrentMemoryContext));
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                             |
      |                             Datum {aka long unsigned int}
/usr/local/packages/postgresql-16.x/include/server/fmgr.h:361:53: note: in definition of macro ‘PG_RETURN_POINTER’
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c:305:7: note: in expansion of macro ‘PG_RETURN_ARRAYTYPE_P’
  305 |       PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate, CurrentMemoryContext));
      |       ^~~~~~~~~~~~~~~~~~~~~
In file included from pgmoneta/pgmoneta_ext/src/include/utils.h:38:
/usr/local/packages/postgresql-16.x/include/server/postgres.h:322:29: note: expected ‘const void *’ but argument is of type ‘Datum’ {aka ‘long unsigned int’}
  322 | PointerGetDatum(const void *X)
      |                 ~~~~~~~~~~~~^
make[2]: *** [src/CMakeFiles/pgmoneta_ext.dir/build.make:76: src/CMakeFiles/pgmoneta_ext.dir/pgmoneta_ext/lib.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:935: src/CMakeFiles/pgmoneta_ext.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

Maybe we need changes to test suite...

Compiled with

```
export CC=/usr/bin/gcc && cmake -DCMAKE_BUILD_TYPE=Debug ..
```

but we may need

```
export CC=/usr/bin/clang && cmake -DCMAKE_BUILD_TYPE=Debug ..
```

as well

**Comment by jesperpedersen** [comment]
Why did you disable the CI ?

**Comment by jesperpedersen** [comment]
Make sure that you are happy with the patch as a whole, and then send a PTAL

**Comment by GuChad369** [comment]
> Why did you disable the CI?

To make testing more convenient. To test our test suite in a container, I need to commit changes to enable the container to download the updated code version. Therefore, I disabled the CI to speed up the testing process.

I noticed it is available for version 17, so I have added PostgreSQL version 17 to our test suite, and now everything works well in the container.

I have also added new tests for our new functions.

Screenshot 2024-08-06 at 10 19 34 AM

There is one issue with the CI: our test suite needs to retrieve a file from the container, which does not exist in the CI environment, causing it to return an error. How should we handle this in the CI?

Screenshot 2024-08-06 at 10 56 11 AM

**Comment by GuChad369** [comment]
Can we remove the test suite from the CI and only call a single function, like `pgmoneta_ext_version`, to ensure it installs successfully?

**Comment by jesperpedersen** [comment]
We can disable the test suite in CI, check if there something that need to be updated for the developer guide

File: doc/manual/user-02-functions.md [hunk 4]
```diff
@@ -13,5 +13,10 @@ After you create the extension `pgmoneta_ext` using the `postgres` role, you can
 | `pgmoneta_ext_version()`    |   Default |    None    | Return the version number of `pgmoneta_ext` as a Datum.|
 | `pgmoneta_ext_switch_wal()` | SUPERUSER |    None    | A function for switching to a new WAL file.            |
 | `pgmoneta_ext_checkpoint()` | SUPERUSER <br>pg_checkpoint | None | A function which forces a checkpoint. <br>This function can only be executed by a `SUPERUSER` in PostgreSQL 13/14, but can also be executed by `pg_checkpoint` in PostgreSQL 15+. <br>You can use the SQL command `GRANT pg_checkpoint TO repl;` to assign the role in PostgreSQL 15+.|
+<<<<<<< HEAD
```
Comment (jesperpedersen) [inline doc/manual/user-02-functions.md]: Whoops

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by GuChad369** [comment]
This new version is all set and can work on all versions from 13 to 17.

If you want to test, please remember to replace line 80 in Dockerfile.rocky9 with:
``` Docker
RUN git clone --branch "extra" --single-branch --depth 1 https://github.com/GuChad369/pgmoneta_ext
```

File: src/pgmoneta_ext/lib.c [hunk 5]
```diff
@@ -195,6 +218,7 @@ pgmoneta_ext_get_oid(PG_FUNCTION_ARGS)
    }
 }
 
+
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: Extra linefeed

File: src/include/utils.h [hunk 6]
```diff
@@ -38,13 +38,19 @@ extern "C" {
 #include <postgres.h>
 #include <fmgr.h>
 #include <nodes/pg_list.h>
+#include <funcapi.h>
+#include <utils/array.h>
 
 /* system */
+#include <dirent.h>
 #include <stdbool.h>
+#include <sys/stat.h>
```
Comment (jesperpedersen) [inline src/include/utils.h:48]: Direct include first, then directory includes after - all sorted

File: src/pgmoneta_ext/lib.c [hunk 7]
```diff
@@ -38,23 +38,46 @@
...
+#include <utils/memutils.h>
 #include <utils/syscache.h>
+#include <storage/fd.h>
 
+/* system */
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <dirent.h>
+
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c:64]: Sorting

File: src/pgmoneta_ext/lib.c [hunk 8]
```diff
@@ -38,23 +38,46 @@
...
+#ifdef PG_MODULE_MAGIC
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: Why `#ifdef` ?

File: src/pgmoneta_ext/lib.c [hunk 9]
```diff
@@ -255,3 +279,178 @@ pgmoneta_ext_get_oids(PG_FUNCTION_ARGS)
...
+      {
+         astate = accumArrayResult(astate, CStringGetTextDatum(file_path), false, TEXTOID, CurrentMemoryContext);
+      }
+      else
+      {
+         ereport(ERROR, errmsg_internal("Argument is neither a file nor a directory: %s", file_path));
+         PG_RETURN_NULL();
+      }
+
+      // PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate, CurrentMemoryContext));
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: Comment

File: src/pgmoneta_ext/utils.c [hunk 10]
```diff
@@ -35,8 +35,8 @@
 #include <access/heapam.h>
 #include <access/htup_details.h>
 #include <catalog/indexing.h>
-#include <catalog/pg_authid.h>
 #include <catalog/pg_database.h>
+#include <catalog/pg_authid.h>
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/utils.c]: Sorting

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/pgmoneta_ext/lib.c (hunk 8 above)
Comment (GuChad369) [reply src/pgmoneta_ext/lib.c]: This got messed up by the rebase.

File: src/pgmoneta_ext/lib.c (hunk 9 above)
Comment (GuChad369) [reply src/pgmoneta_ext/lib.c]: The reason for the version 16 error is a type mismatch. I changed this using a simple method and will retest it.

**Comment by GuChad369** [comment]
This new version works well.

File: src/pgmoneta_ext/lib.c [hunk 11]
```diff
@@ -38,23 +38,44 @@
...
 #include <utils/lsyscache.h>
+#include <utils/memutils.h>
 #include <utils/syscache.h>
+#include <storage/fd.h>
+
+/* system */
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <dirent.h>
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c:60]: Sorting

File: src/pgmoneta_ext/lib.c [hunk 12]
```diff
@@ -255,3 +276,176 @@ pgmoneta_ext_get_oids(PG_FUNCTION_ARGS)
...
+      memcpy(VARDATA(file_data), result.data, result.len);
+
+      // Encode as Base64
+      base64_result = encode_bytea_to_base64(file_data);
+
+      PG_RETURN_TEXT_P(base64_result);
+   }
+   else
+   {
+      ereport(LOG, errmsg_internal("pgmoneta_ext_checkpoint: Current role is not a superuser"));
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: `_get_file`

File: src/pgmoneta_ext/lib.c [hunk 13]
```diff
@@ -255,3 +276,176 @@ pgmoneta_ext_get_oids(PG_FUNCTION_ARGS)
...
+      {
+         ereport(ERROR, errmsg_internal("Argument is neither a file nor a directory: %s", file_path));
+         PG_RETURN_NULL();
+      }
+
+      PG_RETURN_DATUM(makeArrayResult(astate, CurrentMemoryContext));
+   }
+   else
+   {
+      ereport(LOG, errmsg_internal("pgmoneta_ext_checkpoint: Current role is not a superuser"));
```
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: `_get_files`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
Look at your patch, and ask yourself if every single change is correct.

It will make for much faster merges

File: src/pgmoneta_ext/lib.c (hunk 11 above)
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c:60]: Sorting

File: src/pgmoneta_ext/lib.c (hunk 12 above)
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: Still `_checkpoint`

File: src/pgmoneta_ext/lib.c (hunk 13 above)
Comment (jesperpedersen) [inline src/pgmoneta_ext/lib.c]: Again

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by GuChad369** [comment]
This is the newly updated version.

**Comment by jesperpedersen** [comment]
Merged.

Thanks for your contribution !

**Pull request merged**